Simplenote/InterlinkProcessor.swift

```diff
@@ -0,0 +1,137 @@
+            return
+        }
+
+        displayInterlinkWindow(around: keywordRange)
+    }
+
+    /// Dismisses the Interlink Window when ANY of the following evaluates **true**:
+    ///
+    ///     1.  There is Highlighted Text in the editor (or)
+    ///     2.  There is no Interlink `[keyword` at the selected location
+    ///
+    @objc
+    func dismissInterlinkLookupIfNeeded() {
+        guard mustDismissInterlinkLookup else {
+            return
+        }
+
+        dismissInterlinkWindow()
+    }
+}
+
+
+// MARK: - Interlinking Autocomplete: Private API(s)
+//
+private extension InterlinkProcessor {
+
+    /// Indicates if the Selected Range's Length is non zero: at least one character is highlighted
+    ///
+    var isSelectingText: Bool {
+        parentTextView.selectedRange().length != .zero
+    }
+
+    /// Indicates if there's an ongoing Undo Operation in the Text Editor
+    ///
+    var isUndoingEditOP: Bool {
+        parentTextView.undoManager?.isUndoing == true
+    }
+
+    /// Indicates if the Interlink Window is visible
+    ///
+    var isInterlinkWindowOnScreen: Bool {
+        interlinkWindowController.window?.parent != nil
+    }
+}
+
+
+// MARK: - Interlinking Autocomplete: Private API(s)
+//
+private extension InterlinkProcessor {
+
+    /// Indicates if we should process Interlink Lookup
+    ///
+    var mustProcessInterlinkLookup: Bool {
+        isUndoingEditOP == false && isSelectingText == false
+    }
+
+    /// Indicates if we should dismiss the Interlink Window
+    ///
+    var mustDismissInterlinkLookup: Bool {
+        isSelectingText || isInterlinkWindowOnScreen && parentTextView.interlinkKeywordAtSelectedLocation == nil
+    }
+
+    /// Presents the Interlink PopoverWindow at a given Editor Range (Below / Above!)
+    ///
+    func displayInterlinkWindow(around range: Range<String.Index>) {
+        let locationOnScreen = parentTextView.locationOnScreenForText(in: range)
+        let parentWindow = parentTextView.window!
+
+        interlinkWindowController.attach(to: parentWindow)
+        interlinkWindowController.positionWindow(relativeTo: locationOnScreen)
+    }
+
+    /// DIsmisses the Interlink Window
+    ///
+    func dismissInterlinkWindow() {
+        interlinkWindowController.close()
+    }
+
+    /// Refreshes the Interlinks for a given Keyword at the specified Replacement Range (including Markdown `[` opening character).
+    /// - Returns: `true` whenever there *are* interlinks to be presented
+    ///
+    func refreshInterlinks(for keywordText: String, in replacementRange: Range<String.Index>, excluding excludedID: NSManagedObjectID?) -> Bool {
+        interlinkViewController.onInsertInterlink = { [weak self] text in
+            self?.parentTextView.insertTextAndLinkify(text: text, in: replacementRange)
+            self?.dismissInterlinkWindow()
+        }
+
+        return interlinkViewController.refreshInterlinks(for: keywordText, excluding: excludedID)
+    }
+}
```

Comment:
I find this method confusing, because it does quite a few different things :-)
* because it is called `refresh..` and returns boolean. Having 0 results I think is still a success. What do you think about having a different method to check if there are interlinks to display or not? Can call it `isempty` or `hasResults`
* do you think maybe it's worth extracting `onInsertInterlink` listener into separate method? And maybe set it up before or after we show the window, around this line `displayInterlinkWindow(around: keywordRange)`

Reply:
Update: I see you changed this in the next PR 😆 
Please disregard the comment above.